### PR TITLE
changes done in UserDetailsCard.jsx

### PR DESCRIPTION
**Pull request by UdayGoel2000**
required review on some changes

### CODE REVIEW

File: src/components/UsersDetailsCard.jsx [hunk 1]
```diff
@@ -33,10 +33,45 @@ const tableHeaders = [
...
+    selectedArray: [],
+    searchText: "",
+    open: false,
+    selectAll: false,
+    buttonArray: [],
+    selectedButton: "bt_1",
+  });
+
+  function handleOpenClose() {
+    if (!open) setState({ open: true });
```
Comment (rahulchaudhary2244) [inline src/components/UsersDetailsCard.jsx]: Its better to remove extra if else statement.
try using this
```setState( { open: !open } )```

File: src/components/UsersDetailsCard.jsx [hunk 2]
```diff
@@ -33,10 +33,45 @@ const tableHeaders = [
...
+    else setState({ open: false });
+  }
+  function setSelectAll(value) {
```
Comment (rahulchaudhary2244) [inline src/components/UsersDetailsCard.jsx]: no need to create extra setter functions for every variable. Directly pass setState method as a prop to update the values

File: src/components/UsersDetailsCard.jsx [hunk 3]
```diff
@@ -129,7 +157,8 @@ export default function UsersDetailsCard() {
 
   const editUserDetails = (data) => {
     setApiData(apiData.map((ele) => (ele.id === data.id ? data : ele)));
-    handleClose();
+    // handleClose();
```
Comment (rahulchaudhary2244) [inline src/components/UsersDetailsCard.jsx]: remove commented code

File: src/components/UsersDetailsCard.jsx [hunk 4]
```diff
@@ -161,7 +190,8 @@ export default function UsersDetailsCard() {
         usersData={usersData}
         selectedArray={selectedArray}
         setEditData={setEditData}
-        handleOpen={handleOpen}
+        // handleOpen={handleOpen}
```
Comment (rahulchaudhary2244) [inline src/components/UsersDetailsCard.jsx]: remove commented code

File: src/components/UsersDetailsCard.jsx [hunk 5]
```diff
@@ -171,7 +201,8 @@ export default function UsersDetailsCard() {
       />
 
       <UpdateUserModal
-        handleClose={handleClose}
+        // handleClose={handleClose}
```
Comment (rahulchaudhary2244) [inline src/components/UsersDetailsCard.jsx]: remove commented code

File: src/components/UsersDetailsCard.jsx [hunk 6]
```diff
@@ -33,10 +33,45 @@ const tableHeaders = [
...
+function reducer(state, action) {
+  return { ...state, ...action };
+}
+
 export default function UsersDetailsCard() {
   const { enqueueSnackbar } = useSnackbar();
 
-  const [selectedArray, setSelectedArray] = useState([]);
+  const [
+    { selectedArray, searchText, open, selectAll, buttonArray, selectedButton },
```
Comment (rahulchaudhary2244) [inline src/components/UsersDetailsCard.jsx]: give meaning full names, like ```selectedArray``` if it holds  user id in it then its better to name it as ```selectedUserIds```.
same goes with buttonArray, selectedButton.
try to give meaning full names to variables.

File: src/components/UsersDetailsCard.jsx [hunk 7]
```diff
@@ -33,10 +33,45 @@ const tableHeaders = [
   },
 ];
 
+function reducer(state, action) {
```
Comment (rahulchaudhary2244) [inline src/components/UsersDetailsCard.jsx]: no need to create a reducer function in memory. I think it is very small and we can directly pass it as an arguement   to useReducer hook.

for eg:
```
useReducer( (state, newState) => ( {...state, ...newState } ),
 {
    selectedArray: [],
    searchText: "",
    open: false,
    selectAll: false,
    buttonArray: [],
    selectedButton: "bt_1",
  });
  ```

File: src/components/UsersDetailsCard.jsx (hunk 1 above)
Comment (UdayGoel2000) [reply src/components/UsersDetailsCard.jsx]: removed the fucntion it self

File: src/components/UsersDetailsCard.jsx (hunk 2 above)
Comment (UdayGoel2000) [reply src/components/UsersDetailsCard.jsx]: deleted

File: src/components/UsersDetailsCard.jsx (hunk 3 above)
Comment (UdayGoel2000) [reply src/components/UsersDetailsCard.jsx]: removed

File: src/components/UsersDetailsCard.jsx (hunk 4 above)
Comment (UdayGoel2000) [reply src/components/UsersDetailsCard.jsx]: removed

File: src/components/UsersDetailsCard.jsx (hunk 5 above)
Comment (UdayGoel2000) [reply src/components/UsersDetailsCard.jsx]: removed

File: src/components/UsersDetailsCard.jsx (hunk 7 above)
Comment (UdayGoel2000) [reply src/components/UsersDetailsCard.jsx]: done

File: src/components/UsersDetailsCard.jsx (hunk 6 above)
Comment (UdayGoel2000) [reply src/components/UsersDetailsCard.jsx]: done

File: READMEDEFAULT.txt [hunk 8]
```diff
@@ -85,4 +85,7 @@ git push -u origin main
 
 git fetch origin
 git checkout main
-git pull origin main
\ No newline at end of file
+git pull origin main
```
Comment (rahulchaudhary2244) [inline READMEDEFAULT.txt:88]: remove this file, it is not needed.

File: src/components/UsersDetailsCard.jsx [hunk 9]
```diff
@@ -107,61 +122,66 @@ export default function UsersDetailsCard() {
...
-      setSelectAll(false);
+      setState({ selectedUsersIdArray: arr.filter((ele) => ele !== id) });
+      setState({ selectAll: false });
     } else {
       arr.push(id);
-      setSelectedArray(arr);
-      if (arr.length === usersData.length) setSelectAll(true);
-      else setSelectAll(false);
+      setState({ selectedUsersIdArray: arr });
+      if (arr.length === usersData.length) setState({ selectAll: true });
```
Comment (rahulchaudhary2244) [inline src/components/UsersDetailsCard.jsx]: Why extra if else, they can be a single statement.
```
setState({ selectAll: (arr.length === usersData.length) });
```

File: src/components/UsersDetailsCard.jsx [hunk 10]
```diff
@@ -107,61 +122,66 @@ export default function UsersDetailsCard() {
...
   const handleChange = (e) => {
-    setEditData({ ...editData, [e.target.name]: e.target.value });
+    setState({ editData: { ...editData, [e.target.name]: e.target.value } });
   };
 
   const handleOnClick = (button_id) => {
-    setSelectAll(false);
-    setSelectedArray([]);
-    setSelectedButton(button_id);
+    setState({ selectAll: false });
```
Comment (rahulchaudhary2244) [inline src/components/UsersDetailsCard.jsx]: No need to call setState 3 times, try understanding the reducer function logic. We are destructuring oldState & newState.
```
setState({ selectAll: false,  selectedUsersIdArray: [] ,  selectedButtonId: button_id });
```
Make this change wherever needed.

File: src/components/UsersDetailsCard.jsx [hunk 11]
```diff
@@ -107,61 +122,66 @@ export default function UsersDetailsCard() {
...
+        changeValue={(searchedValue) => setState({ searchText: searchedValue })}
+      />
 
       <RecordTable
         usersData={usersData}
-        selectedArray={selectedArray}
-        setEditData={setEditData}
-        handleOpen={handleOpen}
+        selectedUsersIdArray={selectedUsersIdArray}
+        setEditData={(data) => setState({ editData: data })}
```
Comment (rahulchaudhary2244) [inline src/components/UsersDetailsCard.jsx]: try giving editData a meaningful name, like userModalData.

File: src/components/UsersDetailsCard.jsx [hunk 12]
```diff
@@ -67,28 +82,28 @@ export default function UsersDetailsCard() {
     return () => {
       clearTimeout(timer);
     };
-  }, [searchText, apiData, selectedButton, selectedArray, selectAll]);
+  }, [searchText, apiData, selectedButtonId]);
```
Comment (rahulchaudhary2244) [inline src/components/UsersDetailsCard.jsx]: why added selectedButtonId & apiData in dependencies ?

I think this useEffect can be improved. Just try to debug it & remove unwanted code & make this code better in terms of performance & readability. Don't hesitate to create a debounce function in utils file.

File: READMEDEFAULT.txt (hunk 8 above)
Comment (UdayGoel2000) [reply READMEDEFAULT.txt:88]: done

File: src/components/UsersDetailsCard.jsx (hunk 9 above)
Comment (UdayGoel2000) [reply src/components/UsersDetailsCard.jsx]: done

File: src/components/UsersDetailsCard.jsx (hunk 10 above)
Comment (UdayGoel2000) [reply src/components/UsersDetailsCard.jsx]: done

File: src/components/UsersDetailsCard.jsx (hunk 11 above)
Comment (UdayGoel2000) [reply src/components/UsersDetailsCard.jsx]: done

**Pull request merged**